k-distribution/include/kframework/builtin/ffi.md

```diff
@@ -0,0 +1,177 @@
+                  | "#uint64" [klabel(#uint64), symbol]
+                  | "#sint64" [klabel(#sint64), symbol]
+                  | "#float" [klabel(#float), symbol]
+                  | "#double" [klabel(#double), symbol]
+                  | "#uchar" [klabel(#uchar), symbol]
+                  | "#schar" [klabel(#schar), symbol]
+                  | "#ushort" [klabel(#ushort), symbol]
+                  | "#sshort" [klabel(#sshort), symbol]
+                  | "#uint" [klabel(#uint), symbol]
+                  | "#sint" [klabel(#sint), symbol]
+                  | "#ulong" [klabel(#ulong), symbol]
+                  | "#slong" [klabel(#slong), symbol]
+                  | "#longdouble" [klabel(#longdouble), symbol]
+                  | "#struct" "(" List ")" [klabel(#struct), symbol]
+endmodule
+
+module FFI
+  imports FFI-SYNTAX
+  imports BYTES
+  imports STRING
+
+```
+
+FFI Calls
+---------
+
+The `#ffiCall` functions are designed to call a native C ABI function and 
+return a native result. They come in three variants:
+
+### Non-variadic
+
+In the first variant, `#ffiCall(Address, Args, ArgTypes, ReturnType)` takes
+an integer address of a function (which can be obtained from
+`#functionAddress`), a `List` of `Bytes` containing the arguments of the
+function, a `List` of `FFIType`s containing the types of the parameters of the
+function, and an `FFIType` containing the return type of the function, and 
+returns the return value of the function as a `Bytes`.
+
+```k
+  syntax Bytes ::= "#ffiCall" "(" Int "," List "," List "," FFIType ")" [function, hook(FFI.call)]
+```
+
+### Variadic
+
+In the second variant,
+`#ffiCall(Address, Args, FixedTypes, VariadicTypes, ReturnType` takes an
+integer address of a function, a `List` of `Bytes` containing the arguments
+of the call, a `List` of `FFIType`s containing the types of the fixed
+parameters of the function, a `List` of `FFIType`s containing the types of the
+variadic parameters of the function, and an `FFIType` containing the return
+type of the function, and returns the return value of the function as a
+`Bytes`.
+
+```k
+                 | "#ffiCall" "(" Int "," List "," List "," List "," FFIType ")" [function, hook(FFI.call_variadic)]
+```
+
+### Generic
+
+In the third variant,
+`#ffiCall(IsVariadic, Address, Args, ArgTypes, NFixed, ReturnType` takes
+a boolean indicating whether the function is variadic or not, an integer
+address of a function, a `List` of `Bytes` containing the arguments of the
+call, a `List` of `FFIType`s containing the parameter typess of the call
+followed by the types of the variadic arguments of the call, if any, an `Int`
+containing how many of the arguments of the call are fixed or not, and an
+`FFIType` containing the return type of the function, and returns the return
+value of the function as a `Bytes`.
+
+```k
+                 | "#ffiCall" "(" Bool "," Int "," List "," List "," Int "," FFIType ")" [function]
+
```

Comment:
```suggestion
  syntax Bytes ::= "#ffiCall" "(" Bool "," Int "," List "," List "," Int "," FFIType ")" [function]

```

Reply:
Another option would be to just declare all three together, with the documentation above all together.